src/app/cli/src/cli_entrypoint/mina_cli_entrypoint.ml

```diff
@@ -949,8 +953,16 @@ let setup_daemon logger =
                 Mina_user_error.raise
                   "You cannot provide both `block-producer-key` and \
                    `block_production_pubkey`"
-            | None, None ->
-                Deferred.return None
+            | None, None -> (
+                match Sys.getenv "MINA_BP_PRIVKEY" with
+                | Some base58_privkey ->
+                    let kp =
+                      Private_key.of_base58_check_exn base58_privkey
+                      |> Keypair.of_private_key_exn
+                    in
+                    return (Some kp)
+                | None ->
+                    return None )
```

Comment:
I would probably prefer to use `Option.first_some` rather than nest pattern-matching like this. Sadly, the function would not have the same lazy behaviour as this and also it only accepts two possibilities, where we need 3. But it would look a lot nicer. Alternatively, you could check `Sys.getenv` up front, as this is a cheap operation, and pattern-match on 3 options immediately. I just find nested pattern matching hard to read. Feel free to disregard it, if you think the current solution is better.